changefeedccl: write new span-level checkpoint

andyyang890 opened the pull request:
This patch starts writing the new span-level checkpoint on clusters
that are running v25.2 or newer.

Fixes #137692

Release note: None

andyyang890 committed:
changefeedccl: write new span-level checkpoint

This patch starts writing the new span-level checkpoint on clusters
that are running v25.2 or newer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141689)
<!-- Reviewable:end -->

asg0451 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -605,7 +605,7 @@ func (ca *changeAggregator) setupSpansAndFrontier() (spans []roachpb.Span, err e
 
 	// Convert the legacy checkpoint to the new span-level checkpoint so that we
 	// can ignore it from this point on.
-	if !ca.spec.Checkpoint.IsZero() {
```
nice catch

asg0451 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
as discussed consider using the new checkpoint and having a method to convert it to the old representation.

andyyang890 replied:
I made this change in a fixup commit, I'll squash it before merging.

asg0451 reviewed (approved):
some minor nits but looks good!

andyyang890 commented on pkg/ccl/changefeedccl/checkpoint/checkpoint.go:
```diff
@@ -118,7 +113,35 @@ func ConvertLegacyCheckpoint(
+		return nil
+	}
+
+	checkpointGoMap := checkpoint.ToGoMap()
```
I originally did that mostly to make the for loop look a little nicer but now that we've upgraded to 1.23 I've replaced it with an iterator :)

andyyang890 commented on pkg/ccl/changefeedccl/checkpoint/checkpoint_test.go:
```diff
@@ -248,7 +247,69 @@ func TestCheckpointRestore(t *testing.T) {
+		"some spans ahead of frontier": {
+			frontier: ts(1),
+			spans: checkpointSpans{
+				{span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")}, ts: ts(1)},
```
There's already a test case in `TestCheckpointMake` for this (see "adjacent spans ahead of frontier merged before being checkpointed"). I added a test case to `TestCheckpointMakeRestoreRoundTrip` like you requested.

andyyang890 committed:
iterutil: add some generic iterator utility functions

This patch adds the `Enumerate` and `Keys` iterator adapters and the
`MinFunc` helper function, which make working with iterators easier.

Release note: None

andyyang890 commented:
TFTR!

bors r=asg0451
